lib/Model/ConvertFunctionType.cpp

```diff
@@ -0,0 +1,664 @@
+  if (!verify<A>(Function.Arguments, CC::GeneralPurposeArgumentRegisters))
+    return std::nullopt;
+  if (!verify<A>(Function.ReturnValues, CC::GeneralPurposeReturnValueRegisters))
+    return std::nullopt;
+
+  constexpr model::Register::Values PTCRR = CC::ReturnValueLocationRegister;
+  if (PTCRR != model::Register::Invalid)
+    revng_assert(model::Register::getArchitecture(PTCRR) == A);
+  for (auto &SavedRegister : CC::CalleeSavedRegisters)
+    revng_assert(model::Register::getArchitecture(SavedRegister) == A);
+
+  model::CABIFunctionType Result;
+  Result.ABI = V;
+
+  auto ArgumentList = convertArguments<CC>(Function.Arguments,
+                                           CC::GeneralPurposeArgumentRegisters,
+                                           TheBinary);
+  if (ArgumentList == std::nullopt)
+    return std::nullopt;
+  for (auto &Argument : *ArgumentList)
+    Result.Arguments.insert(Argument);
+
+  auto
+    ReturnValue = convertReturnValue<CC>(Function.ReturnValues,
+                                         CC::GeneralPurposeReturnValueRegisters,
+                                         CC::ReturnValueLocationRegister,
+                                         TheBinary);
+  if (ReturnValue == std::nullopt)
+    return std::nullopt;
+  Result.ReturnType = *ReturnValue;
+  return Result;
+}
+
+std::optional<model::CABIFunctionType>
+convertToCABIFunctionType(const model::RawFunctionType &Function,
+                          model::abi::Values ABI,
+                          model::Binary &TheBinary) {
+  revng_assert(ABI != model::abi::Invalid);
+  return skippingEnumSwitch<1>(ABI, [&]<model::abi::Values A>() {
+    return convertToCABIFunctionTypeImpl<A>(Function, TheBinary);
+  });
+}
+
+//
+// toRaw
+//
+
+using IndexType = decltype(model::Argument::Index);
+using RegisterList = llvm::SmallVector<model::Register::Values, 1>;
+struct DistributedArgument {
+  RegisterList Registers = {};
+  size_t Size = 0, SizeOnStack = 0;
+};
+using DistributedArguments = llvm::DenseMap<IndexType, DistributedArgument>;
+
+using ArgumentContainer = SortedVector<model::Argument>;
+template<typename CallConv>
+static std::optional<DistributedArguments>
+distributeArgumentsBasedOnPositions(const ArgumentContainer &Arguments) {
+  DistributedArguments Result;
+
+  for (const model::Argument &Argument : Arguments) {
+    auto &Distributed = Result[Argument.Index];
+
+    auto MaybeSize = Argument.Type.size();
+    if (MaybeSize == std::nullopt)
+      return std::nullopt;
+    Distributed.Size = *MaybeSize;
+
+    if (Argument.Type.isFloat()) {
+      if (Argument.Index < CallConv::VectorArgumentRegisters.size()) {
+        auto Register = CallConv::VectorArgumentRegisters[Argument.Index];
+        Distributed.Registers.emplace_back(Register);
+      } else {
+        Distributed.SizeOnStack = Distributed.Size;
+      }
+    } else {
+      if (Argument.Index < CallConv::GeneralPurposeArgumentRegisters.size()) {
```

Comment:
I'm confused, I guess you shuold be treating aggregates in a special way.
Here you seem to assume that each argument fits in a GPR. Why?

Reply:
This is for position based ABI. Any aggregate always fits into the GPR, mostly as a pointer, but still.

Reply:
So position-based ABIs never unpack a struct into registers? Every aggregate larger than a register is passed by pointer to a variable allocated by the caller on the stack?
I find this very surprising. Can you document this either here, or where you define position-based ABIs?

Reply:
Technically, it should be possible for them to unpack aggregates that fit into a single register, but as far as my search went, I never found any evidence of that happening, neither in docs nor in practical compiler tests. If we ever run into something like that, it can be changed, but for now this should suffice.

Reply:
Gotcha.